src/main/java/com/opentable/kafka/partition/ManualKafkaPartitioningStrategy.java

```diff
@@ -0,0 +1,132 @@
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.AdminClientConfig;
+import org.apache.kafka.clients.admin.DescribeTopicsResult;
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.opentable.service.AppInfo;
+
+/**
+ * Uses a consistent manual partitioning strategy. On Mesos, therefore the same instance # will always be
+ * assigned the same partitions assuming
+ * - the total number of partitions in topic hasn't changed
+ * - the total instance count has not changed.
+ *
+ * NOTE: Normally the right thing to do is to use automatic rebalancing (Kafka default), which is a proven
+ * scalable strategy that elastically scales. Use this class only if there is a reason
+ */
+public class ManualKafkaPartitioningStrategy implements KafkaPartitioningStrategy {
+        private static final Logger LOG = LoggerFactory.getLogger(ManualKafkaPartitioningStrategy.class);
+        private final BrokerConfig brokerConfig;
+        private final AppInfo appinfo;
+        private final ImmutableList<Integer> partitionList;
+
+        @Inject
+        public ManualKafkaPartitioningStrategy(
+                final BrokerConfig brokerConfig,
+                final AppInfo appInfo
+                ) throws ExecutionException, InterruptedException {
+
+            this.brokerConfig = brokerConfig;
+            final int partitionCount = numberOfPartitions(brokerConfig.getTopic(), brokerConfig.getBrokerList());
+            this.appinfo = appInfo;
+            final int instanceNumber = calculateInstanceNumber();
+            final int instanceCount = calculateTotalInstances();
+            // If kafka is enabled and instance > partition probably a muck up.
+            if (((instanceCount > partitionCount) && (brokerConfig.isEnabled()))) {
+                throw new IllegalStateException("More instances than partitions - probably a misconfiguration?");
+            }
+            // Calculate distribution of instances to partitions
+            Multimap<Integer, Integer> map = getInstanceToPartitionMap(partitionCount, instanceCount);
+            this.partitionList = ImmutableList.copyOf(map.get(instanceNumber - 1));
+            LOG.info("Partition count {}, instance number {}, instance count {} partitions to assign {}", partitionCount, instanceNumber, instanceCount, partitionList);
+        }
+
+        @VisibleForTesting
+        Multimap<Integer, Integer> getInstanceToPartitionMap(final int partitionCount, final int instanceCount) {
+            final Multimap<Integer, Integer> map = HashMultimap.create();
+            for (int partition = 0; partition < partitionCount; partition++) {
+                int assignedInstance = partition % instanceCount;
+                map.put(assignedInstance, partition);
+            }
+            return map;
+        }
```

Comment:
This is really the core logic.  I imagine an implementation where you can swap these out.  A "strategy pattern" perhaps?

Reply:
Done